File: _crypto_scan.py
import json, statistics
from urllib.request import urlopen, Request
from urllib.parse import urlencode

def get_json(url, params=None, headers=None, timeout=30):
    if params:
        url = url + ('?' + urlencode(params))
    req = Request(url, headers=headers or {})
    with urlopen(req, timeout=timeout) as r:
        return json.loads(r.read().decode('utf-8'))

# CoinGecko top 100 markets
url='https://api.coingecko.com/api/v3/coins/markets'
params={
 'vs_currency':'usd',
 'order':'market_cap_desc',
 'per_page':100,
 'page':1,
 'sparkline':'false',
 'price_change_percentage':'1h,24h,7d'
}
coins=get_json(url,params=params,headers={'accept':'application/json'})

rows=[]
for c in coins:
    mc=c.get('market_cap') or 0
    vol=c.get('total_volume') or 0
    v_mc=(vol/mc) if mc else 0
    rows.append({
        'id':c.get('id'),
        'name':c.get('name',''),
        'sym':(c.get('symbol','') or '').upper(),
        'rank':c.get('market_cap_rank'),
        'mc':mc,
        'vol':vol,
        'v_mc':v_mc,
        'ch1':c.get('price_change_percentage_1h_in_currency'),
        'ch24':c.get('price_change_percentage_24h_in_currency'),
        'ch7':c.get('price_change_percentage_7d_in_currency'),
    })

v_mcs=[x['v_mc'] for x in rows if x['v_mc']>0]
med=statistics.median(v_mcs) if v_mcs else 0
q90=statistics.quantiles(v_mcs,n=10)[8] if len(v_mcs)>=20 else (max(v_mcs) if v_mcs else 0)

signals=[]
for x in rows:
    ch1=float(x['ch1']) if x['ch1'] is not None else 0.0
    ch24=float(x['ch24']) if x['ch24'] is not None else 0.0
    v=float(x['v_mc'])

    move_score=0
    if abs(ch1)>=6: move_score+=3
    elif abs(ch1)>=4: move_score+=2
    elif abs(ch1)>=2.5: move_score+=1
    if abs(ch24)>=20: move_score+=3
    elif abs(ch24)>=12: move_score+=2
    elif abs(ch24)>=8: move_score+=1

    vol_score=0
    if v>=max(0.25, q90): vol_score+=3
    elif v>=max(0.15, med*2): vol_score+=2
    elif v>=max(0.10, med*1.5): vol_score+=1

    divergence=0
    if vol_score>=2 and abs(ch1)<1.0 and abs(ch24)<3.0:
        divergence=2
    if move_score>=3 and v<med*0.8:
        divergence=max(divergence,1)

    score=move_score+vol_score+divergence
    if score>=5 or (abs(ch1)>=6 and v>=0.08) or (abs(ch24)>=20):
        signals.append((score, x, ch1, ch24, v, move_score, vol_score, divergence))

signals.sort(key=lambda t:(t[0], abs(t[2])+abs(t[3]), t[4]), reverse=True)

# Binance cross-check
binance_map={}
try:
    tickers=get_json('https://api.binance.com/api/v3/ticker/24hr',headers={'accept':'application/json'})
    for t in tickers:
        s=t.get('symbol','')
        if s.endswith('USDT'):
            base=s[:-4]
            if base.endswith(('UP','DOWN','BULL','BEAR')):
                continue
            binance_map[base]=t
except Exception:
    binance_map=None

print('MED_VMC',med,'Q90_VMC',q90,'TOTAL_SIGNALS',len(signals))
for score,x,ch1,ch24,v,ms,vs,dv in signals[:12]:
    base=x['sym']
    bn=binance_map.get(base) if binance_map else None
    bn_pct=float(bn['priceChangePercent']) if bn else None
    bn_qv=float(bn['quoteVolume']) if bn else None
    bn_tr=int(float(bn['count'])) if bn else None
    print(json.dumps({
        'rank':x['rank'],
        'sym':base,
        'name':x['name'],
        'score':score,
        'ch1':ch1,
        'ch24':ch24,
        'ch7':float(x['ch7']) if x['ch7'] is not None else None,
        'v_mc':v,
        'binance': bool(bn),
        'bn24': bn_pct,
        'bnQuoteVol': bn_qv,
        'bnTrades': bn_tr,
        'reason': {'move_score':ms,'vol_score':vs,'divergence':dv}
    }))

File: crypto_scan.js
const fetch = global.fetch;

const cgUrl = 'https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&order=market_cap_desc&per_page=120&page=1&sparkline=false&price_change_percentage=1h,24h';
const stable = new Set([
  'tether','usd-coin','dai','true-usd','frax','paypal-usd','first-digital-usd','usdd','usde','tusd'
]);

function sleep(ms){ return new Promise(r=>setTimeout(r, ms)); }
function pct(a,b){ return (b-a)/a*100; }

(async () => {
  const cg = await (await fetch(cgUrl, { headers: { accept: 'application/json' } })).json();
  const top = cg
    .filter(x => !stable.has(x.id) && !['bitcoin','ethereum'].includes(x.id))
    .slice(0, 100);

  const tickers = await (await fetch('https://api.binance.com/api/v3/ticker/24hr')).json();
  const tmap = new Map(tickers.map(t => [t.symbol, t]));

  const candidates = [];

  for (const c of top) {
    const base = c.symbol.toUpperCase();
    const sym = base + 'USDT';
    if (!tmap.has(sym)) continue;

    const t = tmap.get(sym);
    const change24 = parseFloat(t.priceChangePercent);
    const quoteVol = parseFloat(t.quoteVolume);
    const ch1h = c.price_change_percentage_1h_in_currency;

    // Preselect to reduce API calls
    const preselect = Math.abs(change24) > 6 || quoteVol > 50_000_000 || (typeof ch1h === 'number' && Math.abs(ch1h) > 2);
    if (!preselect) continue;

    await sleep(70);
    const kUrl = `https://api.binance.com/api/v3/klines?symbol=${sym}&interval=1m&limit=120`;
    const kl = await (await fetch(kUrl)).json();
    if (!Array.isArray(kl) || kl.length < 60) continue;

    const closes = kl.map(r => parseFloat(r[4]));
    const vols   = kl.map(r => parseFloat(r[5]));

    const lastPx = closes[closes.length - 1];
    const px15a  = closes[closes.length - 16];
    const move15 = pct(px15a, lastPx);

    const sum = arr => arr.reduce((a,b)=>a+b, 0);
    const vLast30 = sum(vols.slice(-30));
    const vPrev30 = sum(vols.slice(-60, -30));
    const vRatio30 = vPrev30 > 0 ? vLast30 / vPrev30 : null;

    const vLast10 = sum(vols.slice(-10));
    const vPrev10 = sum(vols.slice(-20, -10));
    const vRatio10 = vPrev10 > 0 ? vLast10 / vPrev10 : null;

    const last = kl[kl.length - 1];
    const h = parseFloat(last[2]), l = parseFloat(last[3]), clp = parseFloat(last[4]);
    const rangePct = (h - l) / clp * 100;

    let score = 0;
    if (Math.abs(move15) >= 3) score += 3;
    if (vRatio30 && vRatio30 >= 3) score += 3;
    if (vRatio10 && vRatio10 >= 4) score += 2;
    if (rangePct >= 1.5) score += 1;

    if (score >= 4) {
      candidates.push({ sym, move15, vRatio30, vRatio10, rangePct, change24, quoteVol });
    }
  }

  candidates.sort((a,b) => {
    const av = a.vRatio30 || 1;
    const bv = b.vRatio30 || 1;
    return (Math.abs(b.move15) + Math.log(bv)) - (Math.abs(a.move15) + Math.log(av));
  });

  if (candidates.length === 0) {
    console.log('OK');
    return;
  }

  console.log('ALERTA CRYPTO (últimos ~15-30 min, Binance spot USDT)');
  for (const x of candidates.slice(0, 8)) {
    const dir = x.move15 >= 0 ? '↑' : '↓';
    const vr30 = x.vRatio30 ? x.vRatio30.toFixed(2) + 'x' : 'n/a';
    const vr10 = x.vRatio10 ? x.vRatio10.toFixed(2) + 'x' : 'n/a';
    const qv = x.quoteVol >= 1e9 ? (x.quoteVol/1e9).toFixed(2)+'B'
             : x.quoteVol >= 1e6 ? (x.quoteVol/1e6).toFixed(1)+'M'
             : x.quoteVol.toFixed(0);

    console.log(`- ${x.sym}: ${dir}${x.move15.toFixed(2)}%/15m | Vol 30m ${vr30} (10m ${vr10}) | rango 1m ${x.rangePct.toFixed(2)}% | 24h ${x.change24.toFixed(2)}% | QVol24h ${qv} USDT`);
  }

  console.log('\nCriterio trader (rápido):');
  console.log('- Subida con Vol>=3x: posible continuación; esperar pullback (VWAP/EMA) + stop corto.');
  console.log('- Movimiento fuerte con Vol bajo: más probable fakeout/stop-hunt.');
  console.log('- Caída con Vol alto: riesgo de cascada; evitar longs hasta base clara.');
})().catch(e => {
  console.error('OK');
});
